### ISSUE TITLE
Do not encourage to publish package-lock in the source code version control as a general practice

### ISSUE THREAD

**Issue by cseas** [npm/documentation#810]
Opening this here again as the repository of the earlier reported issue got archived.
https://github.com/npm/npm/issues/20603

The npm documentation mentions it clearly that package-lock is *always* meant to be committed, leaving library authors unaware of the risk that their consumers might install different dependency versions than what they're using. The documentation should be changed to mention that since package-lock files are ignored in published npm packages, library authors should exercise caution when choosing to commit them.

Examples:

1. https://docs.npmjs.com/cli/v6/configuring-npm/package-locks#using-locked-packages
```text
It is highly recommended you commit the generated package lock to source control
```

2. https://docs.npmjs.com/cli/v10/configuring-npm/package-lock-json#description
```text
This file is intended to be committed into source repositories
```

Article with reasoning:
https://gajus.medium.com/stop-using-package-lock-json-or-yarn-lock-909035e94328

cc: @gajus

**Comment by MylesBorins** [comment]
I disagree with this suggestion. While it is true that when a package is a dependent you are going to have a different tree, there is a large benefit to consistent development environments.

Yes, there is inconsistency when people install the package, but just because there are scenarios where there is a lack of consistency doesn't mean there is no value in a development environment that is mostly reproducible. Having consistency when collaborating across teams, reviewing PRs, and debugging 3rd party bug reports removes a ton of randomness and edge cases.

It is definitely possible to have a test run in CI that includes `--package-lock=false` if you want to test what happens when you install w/o the lock, but imho that is far more valuable as a daily task rather than the default scenario. Every tree is going to end up with a different layout / resolution of the tree based on the various other dependencies (prod / dev / peer), so trying to test out for all potential iterations is impossible... what is possible is ensuring you have consistency during development.

I'll leave this open for more discussion, but I would not approve any changes to the current documentation that would remove this recomendation.

**Comment by ljharb** [comment]
I make an .npmrc in all my projects with package-lock=false and i gitignore lockfiles for this reason.

Defaulting a lockfile was the wrong decision in npm 5 and it’s the wrong decision now. “A consistent development environment” is a bug - in the few scenarios where you need reproducibility, you can use `--before to obtain it just as easily as you could use package-lock to disable the lockfile, but the better default is the experience consumers get, even if it’s inconsistent.

**Comment by DanKaplanSES** [comment]
> The documentation should be changed to mention that since package-lock files are ignored in published npm packages, library authors should exercise caution when choosing to commit them to source code.

Wow, I never considered that. In this context, [would a shrinkwrap file be *recommended*](https://docs.npmjs.com/cli/v10/configuring-npm/npm-shrinkwrap-json)? That page says:

> It's strongly discouraged for library authors to publish this [shrinkwrap] file, since that would prevent end users from having control over transitive dependency updates.

**Comment by ljharb** [comment]
Publishing a shrinkwrap file is exceedingly user-hostile, because it prevents end users from getting transitive dependency updates.

In an *app* - something not published - shrinkwrap and package-lock are identical, and should always exist.

**Comment by DanKaplanSES** [comment]
> Publishing a shrinkwrap file is exceedingly user-hostile, because it prevents end users from getting transitive dependency updates.

But it only prevents end-users from getting _that libraries transitive dependencies_, right? That's my interpretation and if it's incorrect I think the documentation should clarify (but that's unrelated to this issue). If I am correct, I don't understand how that can affect the end user, unless the library failed in some other way (e.g., should have used peer dependencies but didn't).

**Comment by ljharb** [comment]
Yes, that's right - it affects the end user because the dependencies *of your library* can't be deduped, or upgraded, which means if there's a bug or security vulnerability, they're stuck with what's in the shrinkwrap file.

**Comment by DanKaplanSES** [comment]
> Yes, that's right - it affects the end user because the dependencies _of your library_ can't be deduped, or upgraded, which means if there's a bug or security vulnerability, they're stuck with what's in the shrinkwrap file.

Ah, gotcha. That makes sense.

**Comment by matthias-ccri** [comment]
Just to chime in here --- the right move is to check in the lockfile always 100% of the time, for all projects: both packages and applications.

> “A consistent development environment” is a bug

I'm sorry but I have to chuckle a bit here. You gotta admit, it is a little funny to say that stability and order is a bug. Sure, you want to catch incompatibilities sooner. But entertain the opposing position for a second --- some developers want to develop without surprises. Maybe stability is a good thing?

Let's also explore what is the easier option:

> where you need reproducibility, you can use `--before` to obtain it just as easily ...

Maybe using `--before` is easy, but it's not _just as easy_ a having a lockfile. You can't commit it or share it as easily. You have to get your CI to run it somehow. I assume we're talking about debugging certain dependency combinations. This is where a lockfile shines, because a lockfile is an easy way to snapshot a dependency tree in time. Create the dependency tree with `--before`, then capture it in a lockfile, and push it to a branch. Golden.

It's actually easy on the other end, if you have a lockfile and want to install bleeding edge deps. Just delete the lockfile and install from scratch. That's just as easy.

Moreover, sorry if this is preachy, but if you want to catch errors, you should do it in an organized way:

1. decide that you want to catch errors
2. install bleeding edge deps and look for errors

The alternative is this:
1. try to do work
2. be surprised by some random breakage
3. be forced to deal with it

Which one is more organized? If npm removes the lockfile by default, it would make users to randomly deal with bleeding edge bugs. Is it worth it?

You also said something that I need to pick at:

> the better default is the experience consumers get

The question is: _which consumers_? 
- Consumers that just created a new project and are getting the bleeding edge versions of all deps?
- Consumers that have existing projects and are adding your package?
- Consumers that have already installed your package at some previous point in time, and are installing or updating other dependencies?

Consumers have lockfiles too, so your consumers can have any variation or combination of deps. Also, consumers don't install your devDependencies, so if you want your library project's node_modules to mirror a consuming project, it's going to be hard. You might as well set up a downstream project just to test the package.

The point is: given the variability in consumer installation scenarios, a lockfile is the best way to capture each scenario in a branch, so it can be developed and tested in an orderly way.

TLDR: lockfiles take the W. Always commit 100% of the time.

**Comment by ljharb** [comment]
@matthias-ccri you can put `before=2024-08-27` in `.npmrc` and commit it, and it'll work in CI too. so yeah, it's just as easy.

**Comment by cseas** [comment]
> package-lock.json is out of sync with package.json. Yes this is needs to be avoided, see 1 above. But ultimately it is a bug in your library. Fix it and publish a new version where they are in sync.

@steinybot Libraries don't publish package-lock files, that's why I created this issue. The documentation doesn't make it clear enough that the users of your library will most likely NOT install the dependencies you're developing your library with. If you consider it a bug, then it's a bug in npm itself, not something library authors can fix.

Quoting the npm documentation itself:

> package-lock.json cannot be published, and it will be ignored if found in any place other than the root project.

Source: https://docs.npmjs.com/cli/v11/configuring-npm/package-lock-json#package-lockjson-vs-npm-shrinkwrapjson

**Comment by matthias-ccri** [comment]
@cseas 

> If we create a package-lock.json, there's a good possibility someone creating one from the same project just 30 minutes later will have a different package-lock.json.

You would **commit and push** the `package-lock.json`, and then it would not change unless 1) you change some `package.json` deps, or 2) someone manually and intentionally regenerates it (which they will then commit and push). Perfect.

> Lockfiles easily cross thousands of lines, that's a lot of versions to expect to remain stable.

The **latest publicly published packages** are not stable, but **the ones you have installed in your project** can be stable ... _if you use a lockfile_. Lockfiles grant stability.

> And updating your package-lock every few hours is the same as just not having one.

When you have a lockfile, you can restore point-in-time node_modules states.

> that won't save the users of their library from the attack

They will be safe because they also use lockfiles, since this is the best practice recommended by _all_ major package managers (npm, yarn, pnpm).

> library authors need to take additional steps to protect users when depending on package-lock files

They should tell their users to use lockfiles. There's no other way to be safe from these supply chain attacks. If the "latest" version of a package is malicious, then it will be installed in any fresh project, when `npm install malicious-package` is run. The only protection is for existing projects to _not_ install "latest", and one great way to do that is to use lockfiles.

> a test run in CI that includes --package-lock=false if you want to test what happens when you install w/o the lock

This is good, but in order for the CI job to properly capture the failing state, it must capture that particular combination of dependencies _in a lockfile_, and then push a branch with that lockfile.

> many developers I've talked to were surprised to find out their users ended up with a different dependency tree than was intended for their library.

This is a problem that's solved by _education_ about the nature of semver and dependencies and the passage of time. Npm is not wrong to suggest that **the lockfile should be committed in all cases**.

I wrote this full explanation of all the reasons why, and rebuttals to all common objections:
https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5/76058921#76058921
